README.md
# bot_ames
A discord bot used for fetching character data in Princess Connect! Re:Dive by CyGames. Ames is planned to provide QoL functions for Clan Battles in the near future. Apart from PCR:D functions, Ames also provides some miscellaneous functions too.

Ames is written in Python 3.7.x utilising the [`rewrite` branch of Discord's Python API](https://github.com/Rapptz/discord.py) and makes use of `cogs` and `extensions` to integrate commands.

## Adding Ames
You can add Ames with [this link](https://discord.com/api/oauth2/authorize?client_id=599290654878597140&permissions=1342565456&scope=bot) or join her development server with [this link](https://discord.gg/fYF8hurRt4). You will need to have the **manage server** permission on the server you wish to add Ames. If Ames does not send anything when she joins, use `.help` to get started.

## Capabilities
Ames' current features:
* Characters
  * Display both EN and JP skill descriptions
  * Show attack pattern
  * Show raw character stats (WIP)
  * Show raw skill stats (WIP)
  * Show character profile
  * Show raw UE stats
* Feed
  * Listen to [PriConne EN twitter](https://twitter.com/priconne_eng) for updates
  * Listen to [PriConne JP twitter](https://twitter.com/priconne_redive) for updates
  * Various other game feeds
* Search
  * Can search characters based on their attributes/skill effects/position to aid team composition
* Gacha
  * Simulate the current gacha banner (JP)

Ames' planned features:
* Clan Battle (JP)
  * Show boss stats and effects
  * Boss reminder system to help with coordination
* Events (JP)
  * Show future events schedule

## Blue Oath JP
As an aside, Ames also provides a small set of Blue Oath (JP) related commands with a semi-standalone module `cog_blueoath.py`, with information and details provided and maintained by the EN Discord community. There are 2 main sources of information; [the spreadsheet](https://docs.google.com/spreadsheets/d/1UrEFf88vjcOFRy7tFOfcedsoobXU7ZTeezFlRFVEo2Q/edit?usp=sharing) and the [EN Wiki](https://blueoath.miraheze.org/wiki/Main_Page). You're welcome to join the [Blue Oath EN discord](https://discord.gg/hzNRN2a) community!

### Blue Oath Capabilities
Use `.bo help` to see all Blue Oath related commands.
* Ships
  * Retrieve ship skills, stats, gallery, traits, limit break information
* Daily Oil reminders

## Disclaimer and Special Thanks
As mentioned already, `Ames` uses a privately maintained database, and additionally cross references data from an unofficial database known as [redive_master_db_diff](https://github.com/esterTion/redive_master_db_diff) and [HatsuneNote's PrincessGuide](https://github.com/superk589/PrincessGuide). Many thanks to their creators for relieving much of the time that would've needed to fetch data.

All art assets used by `Ames` are not my creation and are property of their rightful owners.

## Contact
Find me on discord: `tigertiggs#5376`  
[Join Ames' development server](https://discord.gg/fYF8hurRt4)

changelog
{
    "4.14.12": [
        "+ [phatsune] added flb for misaki, added ue for xyukari, xakino, xsaren",
        "+ [gacha] updated banners"
    ],
    "4.14.11": [
        "+ [phatsune] added x.miyako, x.mahiru",
        "+ [gacha] updated banners",
        "+ [phatsune] updated m.monika's sk1p so it's less vague"
    ],
    "4.14.10": [
        "+ [phatsune] added no.kuuka, added ue for m.tomo, m.monika",
        "+ [gacha] updated banner"
    ],
    "4.14.9": [
        "+ [phatsune] added no.io, added flb for arisa and monika",
        "+ [gacha] updated banners"
    ],
    "4.14.8": [
        "+ [phatsune] added homare",
        "+ [gacha] updated banners"
    ],
    "4.14.7": [
        "+ [phatsune] added karin"
    ],
    "4.14.6": [
        "+ [phatsune] added h.suzuna, added UE for h.rei, h.tsumugi, h.matsuri",
        "+ [gacha] updated banners"
    ],
    "4.14.5": [
        "+ [phatsune] added h.ninon, h.kaori",
        "+ [gacha] updated banners"
    ],
    "4.14.4": [
        "+ [cb] added admin OT adding ability. Thanks zeik."
    ],
    "4.14.3": [
        "+ [phatsune] added credita and note about djeeta flb",
        "+ [gacha] update banners"
    ],
    "4.14.2": [
        "+ [phatsune] added ue for a.akari and a.yori, added o.yuki",
        "+ [gacha] update banners"
    ],
    "4.14.1": [
        "+ [phatsune] add o.mitsuki",
        "+ [gacha] update banner"
    ],
    "4.14.0": [
        "+ [phatsune] added k.shiori/k.hatsune/",
        "+ [phatsune] added ue for neneka",
        "+ [gacha] updated banners",
        "+ [phatsune][pupdate] added new code and configs to support kizuna characters"
    ],
    "4.13.10": [
        "+ [phatsune] added s.chika",
        "+ [phatsune] equipment update"
    ],
    "4.13.9": [
        "+ [phatsune] added ue for s.hatusne, s.jun, s.misato",
        "+ [cb] implemented new green queue"
    ],
    "4.13.8": [
        "+ [phatsune] added aoi flb"
    ],
    "4.13.7":  [
        "+ [phatsune] added s.tsumugi, s.nozomi",
        "+ [gacha] updated banners"
    ],
    "4.13.6": [
        "+ [cb] changed q timeout and displays"
    ],
    "4.13.5": [
        "+ [phatsune] added s.shizuru",
        "+ [gacha] updated banner"
    ],
    "4.13.4": [
        "+ [phatsune] added ue for s.ruka, s.anna, s.nanaka",
        "+ [phatsune] added kaori6",
        "+ [gacha] updated banner"
    ],
    "4.13.3": [
        "+ [phatsune] added s.eriko",
        "+ [gacha] updated banners"
    ],
    "4.13.2": [
        "+ [phatsune] added p.rei",
        "+ [gacha] updated gacha banners"
    ],
    "4.13.1": [
        "+ [ames] updated internal emote json",
        "+ [cb] queue no longer displays entire queue embed if the queue is empty"
    ],
    "4.13.0": [
        "+ [gacha] updated banners",
        "+ [phatsune] updated level caps for characters and ue",
        "+ [phatsune] added f.chloe, tt.inori, tt.kaya, ww.aoi, ww.tamaki, ww.mifuyu, suzume flb, kurumi flb",
        "+ [phatsune] added ue for r.rin, r.mahiru, inori, w.rino, w.ayumi"
    ],
    "4.12.8": [
        "+ [phatsune] added nozomi flb, ue for yuni and chieru",
        "+ [gacha] updated banners"
    ],
    "4.12.7": [
        "+ [gacha] updated spark count",
        "+ [bo] extended the update_tag prevlieges to admins"
    ],
    "4.12.6": [
        "+ [core] added .avatar",
        "- smh roru"
    ],
    "4.12.5": [
        "+ [phatsune] added f.chieru and her assets",
        "+ [phatsune] new prefix school festival (f)"
    ],
    "4.12.4": [
        "+ [core] pinging ames with a command now actually works",
        "+ [phatsune] added cr.maho"
    ],
    "4.12.3": [
        "+ [phatsune] updated for new eq and lvl cap up",
        "+ [gacha] updated banner",
        "- [twitter] added traceback to investigate odd response codes and listener behaviour",
        "- [twitter] forced include_replies attr to be false"
    ],
    "4.12.2": [
        "+ [phatsune] updated ue for imas cg characters"
    ],
    "4.12.1": [
        "+ [phatsune] added lyricals flb",
        "- [twitter] cache limit increased to 100",
        "+ [phatsune] fixed typo in cr.makoto's ub tl",
        "- [pupdate] fixed bug that kept increasing atkptn"
    ],
    "4.12.0": [
        "+ [phatsune] added cr.makoto",
        "- [phatsune] fixed a few tags",
        "+ [pupdate][phatsune] updated structure to accomodate previously missed skills on alt characters, added support for multiple attack patterns",
        "+ [twitter] made ann embed layout nicer"
    ],
    "4.11.2": [
        "+ [phatsune] added pkyaru, crrima",
        "+ [gacha] updated gacha banner"
    ],
    "4.11.1": [
        "- [cb] fixed boss input bounds and added timeout function for queue display"
    ],
    "4.11.0": [
        "+ [cb] added queueing command and its assets",
        "- [cl] fixed changelog",
        "+ [phatsune] added magical prefix alias"
    ],
    "4.10.11": [
        "+ [phatsune] added skasumi, updated gacha banner"
    ],
    "4.10.10": [
        "+ [gacha] added new pages to banner command that shows currently standard characters",
        "- [help] fixed broken md highlights",
        "+ [shenp] coom"
    ],
    "4.10.9": [ 
        "+ [phatsune] adjusted stats for lv cap up",

re_ames.py
# ReReAmes bot
# author: tigertiggs#5376
# 2nd rewrite for better expansion, automation and future proofing
# rewrite will try to add more comments should I drop the project

# dependencies
import datetime, time
import os, sys, traceback
import asyncio, random, json
from difflib import SequenceMatcher as sm
import discord
#from discord.ext.commands import Bot
from discord.ext import commands
#import mysql.connector
#from mysql.connector import pooling
#from mysql.connector import errorcode

# add ames_bot folder to the search path
dir_path = os.path.dirname(os.path.realpath(__file__))
sys.path.insert(1, dir_path+'/commands')

# ames prefix (do not touch)    
BOT_PREFIX = (".")

start_time = None

# utility classes
class logger:
    def __init__(self, client):
        self.logc = None
        self.client = client
        self.private = client.private['ames_logger']
    
    async def send(self, *msg):
        if self.logc == None:   
            self.logc = self.client.get_guild(self.private['guild_id']).get_channel(self.private['channel_id'])
        try:
            await self.logc.send(" ".join([str(w) for w in msg]))
        except:
            pass

# for custom prefixes in different environments
def _prefix(client, message):
    if message.guild:
        if message.guild.id == 202403448694505473:
            return ["a.", BOT_PREFIX]
    return BOT_PREFIX

# main
class Ames(commands.AutoShardedBot):
    def __init__(self):
        global BOT_PREFIX

        self.name = "[Ames]"
        self.dir = dir_path

        # load configs
        print(self.name, "starting")
        with open("commands/_config/config.json") as cf:
            self.config = json.load(cf)
        
        with open(os.path.join(self.config["blueoath_path"], "config.json")) as bocf:
            self.blueoath_config = json.load(bocf)
        
        # override config
        if os.path.exists(os.path.join(self.dir, self.config['override'])):
            with open(os.path.join(self.dir, self.config['override'])) as override:
                print("Found config override - overriding")
                for key, value in list(json.load(override).items()):
                    self.config[key] = value
        
        BOT_PREFIX = self.config['prefix']
        self.prefix = BOT_PREFIX

        # intents
        intent = discord.Intents(
            guilds=True,
            members=True,
            emojis=True,
            messages=True, # guild_messages + dm_messages
            reactions=True, # guild_reactoins + dm_reactions
            typing=True # guild_typing + dm_typing
        )
        #intent.value = 1342565456

        super().__init__(
            command_prefix= _prefix,
            description=    None,
            help_command=   None,
            intents=        intent
        )
        self.command_status = self.config['command_status']

        with open(self.config['emote_path']) as ef:
            self.emotes = json.load(ef)

        with open(self.config['private_path']) as pf:
            self.private = json.load(pf)

        with open('version') as vf:
            self.version = json.load(vf)
        
        # check bot_ames folder tree hierarchy
        self.init_success = self.__check_init()
        if not self.init_success:
            print('failure during initialisation - exiting')
            return

        # load permissions
        with open(self.config['perms_path']) as pf:
            self.perms = json.load(pf)
        
        #with open(self.config['guild_perms_path']) as gpf:
        #    self.guild_perms = json.load(gpf)

        # set logger and db
        self.log = logger(self)
        #self.database = database(self)
        self.database = None

        # load cogs
        self.cogs_status = dict()
        for extension in self.config['command_cogs']:
            try:
                print('loading',extension,end="")
                self.load_extension(extension)
                print('\tsuccess!',flush=True)
            except Exception as err:
                print('\tfailed!\n{0} failed to load:'.format(str(extension)),err,flush=True)
                traceback.print_exc()
                self.cogs_status[str(extension)] = False
            else:
                self.cogs_status[str(extension)] = True

    def __check_init(self):
        print(self.name, "checking critical folders and files")

        for path in self.config['required_folders']:
            print(self.name, "checking folder", path, "...", end="")

            full_path = os.path.join(dir_path, path)

            if not os.path.exists(full_path):
                print("\n", self.name, "creating", path, "...")

                try:
                    os.makedirs(full_path)
                except:
                    traceback.print_exc()
                    return False
            
            print("success", flush=True)
        
        for path in self.config['required_files']:
            print(self.name, "checking file", path, "...", end="")
        
            full_path = os.path.join(dir_path, path)

            if not os.path.exists(full_path):
                print("\n", self.name, "creating", path, "...", end="")

                try:
                    with open(full_path, "w+") as rf:
                        rf.write("{}")
                except:
                    traceback.print_exc()
                    return False 
            
            print("success", flush=True)

        return True

    # handy function to split list l into chunks of length n
    def chunks(self, l, n):
        # For item i in a range that is a length of l,
        for i in range(0, len(l), n):
            # Create an index range for l of n items:
            yield l[i:i+n] 

    # for critical functions
    def _check_author(self, user, mode="default"):
        if user.id == 235361069202145280:
            return True

        try:
            with open(os.path.join(self.dir, self.config['guild_perms_path'], f"{user.guild.id}.json")) as pf:
                perms = json.load(pf)
        except Exception as e:
            print(e)
            perms = dict()

        if mode == "admin":
            # server owner always have admin perm
            if user.guild.owner == user:
                return True

            #print(perms.get("admin", None))
            #print([role.id for role in user.roles])
            #print(perms.get("admin", None) in [role.id for role in user.roles])

            return perms.get("admin", None) in [role.id for role in user.roles]
        else:
            return False

    # load emote resources from servers
    def _load_resource(self):
        self.team = dict()
        self.res_servers = self.private['resource_servers']
        for server in self.res_servers:
            for emj in super().get_guild(server).emojis:
                self.team[emj.name] = f"<:{emj.name}:{emj.id}>"

    # set looping activity messages
    async def st(self):
        switchtime =    60*60
        playing =       0
        #streaming =     1
        #listening =     2
        #watching =      3
        default =       discord.Activity(name="Use .help",      type=discord.ActivityType(playing))
        act_list = [
            discord.Activity(name='with Hatsune',               type=discord.ActivityType(playing)),
            discord.Activity(name='with gacha rates',           type=discord.ActivityType(playing)),
            discord.Activity(name='PrincessConnectReDive',      type=discord.ActivityType(playing))
        ]
        while True:
            await self.change_presence(activity=default)
            await asyncio.sleep(switchtime)
            await self.change_presence(activity=random.choice(act_list))
            await asyncio.sleep(switchtime)

    async def on_ready(self):
        global start_time
        # uptime
        self.init_time = datetime.datetime.utcnow()
        self.s_time = start_time

        # connect the database and grab resources
        #await self.database.connect()
        self._load_resource()

        print(f'Ready: {self.user} (ID: {self.user.id})')
        if self.config['debug'] == 1: print('----Ames is in debug mode----')

        await self.log.send(self.name,'I\'m back!')
        self.loop.create_task(self.st())

    # welcome message when joining a guild
    async def on_guild_join(self, guild):
        await self.log.send(self.name, 'joined', guild.name)
        if guild.id == 598450517253029888:
            hello = self.emotes['sarenh']+'Thanks for having me here!\nMy prefix is `.` (but all blueoath commands will start with `.bo`) - Please use `.bo help` to get started!'
        else:
            hello = self.emotes['sarenh']+'Thanks for having me here!\nMy prefix is `.` - Please use `.help` to get started!'
        general = discord.utils.find(lambda x: x.name == 'general', guild.text_channels)
        if general != None and general.permissions_for(guild.me).send_messages:
            await general.send(hello)
            return
        else:
            for channel in guild.text_channels:
                if channel.permissions_for(guild.me).send_messages:
                    await channel.send(hello)
                    return
        await self.log.send(self.name,'no available channel found!')
    
    # welcome message when a member joins a guild
    async def on_member_join(self, member):
        guild = member.guild

        # attempt to fetch welcome config
        try:
            with open(os.path.join(self.dir, self.config['guild_perms_path'], f"{guild.id}.json")) as gcf:
                guild_welcome = json.load(gcf).get("welcome", None)
        except:
            guild_welcome = None
        
        if not guild_welcome:
            return
        elif not guild_welcome.get("active", False):
            return
        elif not guild_welcome.get("channel", None):
            return
        elif not guild.get_channel(guild_welcome["channel"]):
            return
        else:
            ch = guild.get_channel(guild_welcome["channel"])
            if guild.id == 598450517253029888:
                text = f"Welcome to **{guild.name}**, <@!{member.id}>! I can fetch useful ship skills and stats from the Blue Oath EN Wiki, you can see my commands with `.bo help`\n(please) READ THE PINS, READ THE SHEETS, READ THE WIKI -Ladios"
            else:
                text = f"Welcome to **{guild.name}**, <@!{member.id}>! I'm a PCRD utility bot and you can see my commands with `.help`"
            
            await ch.send(text)

    async def process_commands(self, message):
        ctx = await self.get_context(message)
        #
        # currently empty put is availble for expansion
        #

        if self.config['debug'] == 1:
            if self._check_author(message.author):
                await self.log.send(str(ctx.command))
                await self.invoke(ctx)
            elif ctx.command != None:
                msg = await message.channel.send(f"Ames is currently in debug mode for MT purposes and will not be accepting any commands {self.emotes['ames']}\nThis message will try to delete itself in `10s`")
                await asyncio.sleep(10)
                try:
                    await msg.delete()
                except:
                    pass

        else:
            try:
                #if ctx.command != None:
                #    await self.log.send('[{0.user.name}] `{1}` `{2.channel.guild.name}` `{2.channel.name}` `{2.author.name}` `{2.content}`'.format(
                #        self, datetime.datetime.now(), message))

                # add BO server restriction
                #if self._check_author(message.author):
                #    pass
                if ctx.message.guild: # textchannel/DMchannel
                    if str(ctx.message.guild.id) in self.blueoath_config['restricted_servers'] and ctx.command:
                        if not ctx.command.cog.qualified_name in self.blueoath_config['ames_allowed_cogs'] and not self.config['debug'] and not ctx.message.channel.id in self.blueoath_config['restricted_servers'][str(ctx.message.guild.id)]['exceptions']:
                            msg = await message.channel.send(f"This command is currently not available to the Blue Oath server. See `.bo help` for available functions. {self.emotes['ames']}\nThis message will try to delete itself in `10s`")
                            await asyncio.sleep(10)
                            try:
                                await msg.delete()
                            except:
                                pass
                            finally:
                                return
                if ctx.command: await self.log.send(self.name, ctx.message.channel.guild, f"`{ctx.message.content}`")
                await self.invoke(ctx)
            except Exception as e:
                await self.log.send(self.name, 'failed to process command', e)
            finally:
                pass
                #await ctx.release()

    # command listener
    async def on_message(self,message):
        pref = await self.get_prefix(message)
        if not isinstance(pref, list):
            pref = [pref]

        if not message.author.bot and any([message.content.startswith(p) for p in pref]):
            if self.config['debug'] == 1 and self._check_author(message.author):
                if message.guild:
                    await self.log.send('`[DEBUG MODE]` [{0.user.name}] `{1}` `{2.channel.guild.name}` `{2.channel.name}` `{2.author.name}` `{2.content}`'.format(
                        self, datetime.datetime.now(), message))
                await self.process_commands(message)
            else:
                await self.process_commands(message)

    # error handling
    async def on_command_error(self, ctx, error):
        #if debug == False:
        ignored = (commands.CommandNotFound, commands.UserInputError)
        error = getattr(error, 'original', error)
        #print(error)
        #_errlog(ctx, str(error), __log)
        if isinstance(error, ignored):
            return
        elif isinstance(error, commands.DisabledCommand):
            await ctx.channel.send(f"I\'m currently taking a break from this command {self.emotes['dead']} `(Command Disabled)`")
            await self.log.send(f"Disabled command: {ctx.command.name}")
            return
        elif isinstance(error, commands.CommandOnCooldown):
            msg = await ctx.channel.send(f"This command is on cd - reusuable in `{round(error.retry_after,1)}s`")
            await asyncio.sleep(error.retry_after)
            await msg.delete()
            return
        elif isinstance(error, discord.Forbidden):
            try:
                await ctx.channel.send(f"{self.emotes['ames']} I don\'t have the required permission(s) to perform/finish this command `(Missing Permissions)`")
            except Exception as e:
                await self.log.send("Failed to send error alert", ''.join(traceback.format_exception(type(e), e, e.__traceback__)))
            await self.log.send(f"Ames is missing permissions in: {ctx.command.name}")
        else:
            await self.log.send('<@235361069202145280> command error in', ctx.command.name, f"```prolog\n{''.join(traceback.format_exception(type(error), error, error.__traceback__))}```")

    async def close(self):
        #await super().session.close()
        #self.database.db_pointer.end()
        await self.log.send(self.name, 'shutting down')
        if self.config['save_config_on_exit'] == 1:
            await self.log.send(self.name, 'saving config')
            self.config['command_status'] = self.command_status
            try:
                with open('commands/_config/amesconfig.json', 'w') as config:
                    config.write(json.dumps(self.config, indent=4))
            except Exception as e:
                await self.log.send('failed', e)
            else:
                await self.log.send('success')

        #with open(self.config['private_path'], "w") as p:
        #    p.write(json.dumps(self.private,indent=4))
            
        await super().close()

    def run(self):
        if self.init_success:
            try:
                super().run(self.private['ames_core']['token'], bot=True, reconnect=True)
            except:
                traceback.print_exc()
    
    def error(self):
        error_msg = dict()
        error_msg['inactive'] = 'This command is currently disabled '+self.emotes['maki']
        return error_msg

    # utility classes
    class page_controller():
        def __init__(self, client, make_embed_func, data, chunk_length, index=False):
            self.arrows =   ['⬅','➡']
            self.make =     make_embed_func
            self.data =     data
            self.chunk_l =  chunk_length
            self.index =    index
            self.client =   client
            
            self.pages = []

            self.make_pages()
        
        def make_pages(self):
            chunk = list(self.client.chunks(self.data, self.chunk_l))
            total = len(chunk)

            for i, block in enumerate(chunk):
                if self.index:
                    self.pages.append(self.make(block, (i+1,total)))
                else:
                    self.pages.append(self.make(block))
        
        def start(self):
            return self.pages[0]
        
        def flip(self, mode='r'):
            if mode == 'r':
                self.pages = self.pages[1:] + [self.pages[0]]
            else:
                self.pages = [self.pages[-1]] + self.pages[:-1]
            
            return self.pages[0]

    # utility functions
    async def find_user(self, guild, user:str):
        members = guild.members

        # check if user is a discord id
        try:
            for tk in [">", "<", "@", "!"]:
                user = user.replace(tk,'')
            #print(user)
            user = guild.get_member(int(user))
        except Exception as e:
            await self.log.send(self.name, 'id mismatch', e)
        else:
            return user
        
        # do string match
        # search name
        cutoff = 0.3
        user = user.lower()
        fname = list(filter(lambda x: sm(None, user, x.name.lower(), None).ratio() >= cutoff and user in x.name.lower(), members))
        fnick = list(filter(lambda x: sm(None, user, x.nick.lower() if x.nick != None else '', None).ratio() >= cutoff and user in x.nick.lower(), members))
        #print(user, fname, fnick)
        if len(fname) != 0 and len(fnick) != 0:
            #print(fname[0].name, fnick[0].nick)
            a = sm(None, user, fname[0].name, None).ratio()
            b = sm(None, user, fnick[0].nick, None).ratio()
            if a == b:
                return fnick[0]
            elif a > b:
                return fname[0]
            else:
                return fnick[0]
        elif len(fname) != 0:
            #print(fname[0].name)
            return fname[0]
        elif len(fnick) != 0:
            #print(fnick[0].nick)
            return fnick[0]
        else:
            return None    

    def get_full_name(self, target):
        with open(os.path.join(self.dir, self.config['hatsune_config_path']), encoding='utf-8') as hcf:
            hconfig = json.load(hcf)
        if len(target) > 1:
            if target[1].isupper():
                prefix = hconfig['prefix_title'].get(target[0].lower(), '???')
                return f"{target[1:]} ({prefix})"
        return target
    
    def get_full_name_kai(self, name, prefix=None, short=False):
        with open(os.path.join(self.dir, self.config['hatsune_config_path']), encoding='utf-8') as hcf:
            hconfig = json.load(hcf)
        if prefix:  
            prefix = hconfig['prefix_title'].get(prefix, '???') if not short else hconfig['prefix_shortened'].get(prefix, hconfig['prefix_title'].get(prefix, '???'))
            return f"{hconfig['kizuna_name'].get(name, name.title())} ({prefix})"
        return hconfig['kizuna_name'].get(name, name.title())

if __name__ == "__main__":
    #global start_time
    start_time = time.time()
    AmesBot = Ames()
    AmesBot.run()